mono/profiler/mono-profiler-log.c

```diff
@@ -652,29 +729,31 @@ buffer_unlock (void)
 {
 	mono_memory_barrier ();
 
+	gint32 state = InterlockedRead (&buffer_lock_state);
+
 	// See the comment in buffer_lock ().
-	if (InterlockedReadPointer (&buffer_rwlock_exclusive) == (gpointer) thread_id ())
+	if (state == PROF_TLS_GET ()->small_id << 16)
 		return;
 
-	g_assert (InterlockedRead (&buffer_rwlock_count) && "Why are we trying to decrement a zero reader count?");
+	g_assert (state && "Why are we decrementing a zero reader count?");
+	g_assert (!(state >> 16) && "Why is the exclusive lock held?");
 
-	InterlockedDecrement (&buffer_rwlock_count);
+	InterlockedDecrement (&buffer_lock_state);
 }
 
 // Cannot be used recursively.
 static void
 buffer_lock_excl (void)
 {
-	gpointer tid = (gpointer) thread_id ();
+	gint32 new_ = PROF_TLS_GET ()->small_id << 16;
 
-	g_assert (InterlockedReadPointer (&buffer_rwlock_exclusive) != tid && "Why are we taking the exclusive lock twice?");
+	g_assert (InterlockedRead (&buffer_lock_state) != new_ && "Why are we taking the exclusive lock twice?");
 
-	MONO_ENTER_GC_SAFE;
+	InterlockedWrite (&buffer_lock_exclusive_intent, 1);
```

Comment:
Most OS-provided R/W locks don't provide strong enough guarantees about reader/writer bias, nor do they even let you configure it. It's kind of crazy, and means you're better off rolling your own R/W lock most times that you need one.

In our case, we need writer bias for two reasons:

1. We take the writer lock when initiating STW. So if one or more threads are just sitting and rapidly generating non-GC-related profiler events with zero GC allocations, it could cause massive delay in the whole STW process, which in turn slows the entire app down for no good reason.
2. We take the writer lock from the profiler's helper thread to generate periodic sync points in the log. When a reader is processing a log, it's expected to queue up events between sync points, sort them by timestamp, and then process them. So, we want these to arrive in a timely fashion so that a reader doesn't end up with tens (or even hundreds) of thousands of events between two sync points.

As far as I know, most OS R/W locks are also way heavier in general than what we have here. For example, they need to support recursive read locks, which is very hard (impossible?) to do without either being inefficient or sacrificing writer bias.

(This reminds me that I forgot to update the comments in this PR: The reader lock is not actually recursive.)